docs/madr/decisions/049-meshexternalservice-api.md

```diff
@@ -0,0 +1,518 @@
+
+Technical Story: https://github.com/kumahq/kuma/issues/6330
+
+## Context and Problem Statement
+
+The current implementation of `ExternalService` has significant limitations and warrants a complete redesign.
+
+* Lack of support for * subdomains
+* Lack of support for thick clients (think Kafka or C* where discovery is done as part of the client).
+* Need for pluggability (for example lambda support or equivalent)
+* Working well with gateways (what if your ExternalService relies on SNI/host header to route through a gateway)
+* Applying policies with and without egress
+
+
+## Considered Options
+
+* Creating MeshExternalService + HostnameGenerator
+* Creating MeshExternalService
+* Creating MeshExternalService and using HostnameGenerator for InternalVIP
+
+## Decision Outcome
+
+Chosen option: "Creating MeshExternalService".
+
+### Positive Consequences
+
+* Clearer structure.
+* Enhanced functionalities.
+* Better policy matching capabilities.
+* No need for a HostnameGenerator
+
+### Negative Consequences
+
+* Whole new resource with a lot of new code.
+* Temporarily increased complexity of a product until the migration is done.
+* Problem with supporting Universal without transparent proxy, because one MeshExternalService points to many real domains
+
+## Pros and Cons of the Options
+
+### Creating MeshExternalService
+
+We're introducing a new object called `MeshExternalService`. This object will be a namespace resource and will only be creatable within the `kuma-system` namespace. It can be created within a specific zone, but in this case, it will only be accessible within the zone where it was created.
+
+```yaml
+kind: MeshExternalService
+metadata:
+  name: example
+  namespace: kuma-system
+  labels:
+    kuma.io/mesh: default
+    kuma.io/zone: east-1
+spec:
+  match:
+  - type: InternalVIP # Kuma will generate a domain
+    value: example.ext.svc.local
+    port: 80
+    protocol: http
+  - type: Domain # Existing domain
+    value: httpbin.com
+    port: 80
+    protocol: http
+  - type: CIDR
+    value: 10.1.1.0/24
+    port: 80
+    protocol: http
+  - type: IP
+    value: 192.168.0.1
+    port: 80
+    protocol: http
+  destination:
+    type: Managed # Managed|Passthrough|Extension
+    extension:
+      type: Lambda 
+      config: # type JSON
+        arn: arn:aws:lambda:us-west-2:123456789012:function:my-function
+    endpoints:
+    - address: 1.1.1.1
+      port: 12345
+    - address: example.com
+    - address: unix://....
+    tls:
+      enabled: true
+      version:
+        min: TLSv1_2 # or TLS_AUTO, TLSv1_0, TLSv1_1, TLSv1_2, TLSv1_3
+        max: TLSv1_3 # or TLS_AUTO, TLSv1_0, TLSv1_1, TLSv1_2, TLSv1_3
+      allowRenegotiation: false
+      verification:
+        skipSAN: true # if this is true then subjectAltNames don't take effect
+        subjectAltNames: # if subjectAltNames is not defined then take domain or ips
+          - example.com
+          - "spiffe://example.local/ns/local"
+        caCert: 
```

Comment:
Do we need specific support for using systemCA?

Reply:
By default, when TLS is enabled and caCert is not provided we are going to check in `kuma-dp` default paths where are certificates. When we find one, we are going to send the information in a bootstrapRequest for a config generation. https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/security/ssl#example-configuration